satellite/metainfo/loop.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package metainfo
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+var (
+	// LoopError is a standard error class for this component
+	LoopError = errs.Class("metainfo loop error")
+)
+
+// Observer is an interface defining an observer that can subscribe to the metainfo loop
+type Observer interface {
+	RemoteSegment(context.Context, storj.Path, *pb.Pointer) error
+	RemoteObject(context.Context, storj.Path, *pb.Pointer) error
+	InlineSegment(context.Context, storj.Path, *pb.Pointer) error
+}
+
+type observerContext struct {
+	Observer
+	ctx  context.Context
+	done chan error
+}
+
+func (observer *observerContext) HandleError(err error) bool {
+	if err != nil {
+		observer.done <- err
+		close(observer.done)
+		return true
+	}
+	return false
+}
+
+func (observer *observerContext) Finish() {
+	close(observer.done)
+}
+
+func (observer *observerContext) Wait() error {
+	return <-observer.done
+}
+
+// LoopConfig contains configurable values for the metainfo loop
+type LoopConfig struct {
+	CoalesceDuration time.Duration `help:"how long to wait for new observers before starting iteration" releaseDefault:"5s" devDefault:"5s"`
+}
+
+// Loop is a metainfo loop service
+type Loop struct {
+	config   LoopConfig
+	metainfo *Service
+	join     chan *observerContext
+	done     chan struct{}
+}
+
+// NewLoop creates a new metainfo loop service
+func NewLoop(config LoopConfig, metainfo *Service) *Loop {
+	return &Loop{
+		metainfo: metainfo,
+		config:   config,
+		join:     make(chan *observerContext),
+		done:     make(chan struct{}),
+	}
+}
+
+// Run starts the looping service
+func (service *Loop) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	defer close(service.done)
+
+	for {
+		err := service.runOnce(ctx)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// Join will join the looper for one full cycle until completion and then returns.
+// On ctx cancel the observer will return without completely finishing.
+// Only on full complete iteration it will return nil.
+func (service *Loop) Join(ctx context.Context, observer Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	obsContext := &observerContext{
+		Observer: observer,
+		ctx:      ctx,
+		done:     make(chan error),
+	}
+
+	select {
+	case <-service.done:
+		return context.Canceled
+	case service.join <- obsContext:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return obsContext.Wait()
+}
+
+func (service *Loop) runOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var observers []*observerContext
+
+	defer func() {
+		for _, observer := range observers {
+			observer.Finish()
+		}
+	}()
+
+	select {
```

Comment:
It will make more sense once we add comments, but this `select` is needed to accept the first observer. The `for` loop that follows is on a time limit (`CoalesceDuration`), so it will only accept new observers within a time window after we get the first observer. If we take this `select` out, we will end up doing work that we don't need to following the `for` loop if no observers join in the time limit

Reply:
There’s actually a worse downside, which is that the coalescing time won’t end up being the time specified.

Reply:
I think that `Join` method should specify this constraint, so one can call `Join` but it will have to wait to some `Run` call that can take it, it may be the first one but it may not be.

I'm not clear about this behavior reading the description of the `CoalesceDuration`, `new observers`, in my opinion, doesn't mean that they are all except the first one in the queue.

Reply:
I suspect that it would make the Join/Run more complicated to follow because now you have to coordinate multiple goroutines to get the work, cancellation and errors.

e.g. let's say the first observer starts the Run, then 10 others join it, now the first observer errors and it should exit... this means it would now need to select another observer to take over the work.

Reply:
I think that you misunderstood my comment.
I was only saying that the `Join` should comment the behavior in its doc comment and `CoalesceDuration` help message doesn't suggest to me saying `new observers` that all the observers are new except the first one in the queue.

I didn't suggest any change in the current implementation, I only asked for clarification in the docs.